android/app/src/main/java/io/peggyjo/peggymeter/PDKListener.java

```diff
@@ -2,17 +2,75 @@
 
 import android.content.Context;
 import android.os.Bundle;
+import android.support.annotation.NonNull;
 import android.util.Log;
-
 import com.audacious_software.passive_data_kit.generators.Generators;
 
-public class PDKListener implements Generators.GeneratorUpdatedListener {
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.firestore.DocumentReference;
+import com.google.firebase.firestore.FirebaseFirestore;
+import com.google.firebase.firestore.WriteBatch;
+
+
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+
+public class PDKListener implements Generators.GeneratorUpdatedListener,
+        OnCompleteListener<Void> {
+    private final HashMap<String, Long> timestamps = new HashMap<>();
+    private final FirebaseFirestore db = FirebaseFirestore.getInstance();
+    private final String uid = FirebaseAuth.getInstance().getCurrentUser().getUid();
+
     public void onGeneratorUpdated(String identifier, long timestamp, Bundle data) {
-        Log.i("PDK listener", "id: " + identifier + " | ts: " + timestamp + " | data.size = " + data.size());
-        // TODO: implement uploading to Firebase.
+
+        if (!timestamps.containsKey(identifier)) {
+            String docKey = uid + "-" + identifier + "-" + timestamp;
```

Comment:
This piece of code is duplicated.
Can you restructure your code so that

boolean shouldUpload = shouldUploadData(...);
if (shouldUpload) {
    // upload data
}

Reply:
Yes.